src/visitable.cpp

```diff
@@ -72,14 +72,30 @@ bool visitable<T>::has_item_with( const std::function<bool( const item & )> &fil
     } ) == VisitResponse::ABORT;
 }
 
+/** Sums the two terms, being careful to not trigger overflow.
+ * Doesn't handle underflow.
+ *
+ * @param a The first addend.
+ * @param b The second addend.
+ * @return the sum of the addends, but truncated to std::numeric_limits<int>::max().
+ */
+template <typename T>
+static T sum_no_wrap( T a, T b ) {
+    if( a > std::numeric_limits<T>::max() - b ||
+        b > std::numeric_limits<T>::max() - a ) {
+        return std::numeric_limits<T>::max();
+    }
+    return a + b;
+}
+
 template <typename T>
 static int has_quality_internal( const T& self, const quality_id &qual, int level, int limit )
 {
     int qty = 0;
 
     self.visit_items( [&qual, level, &limit, &qty]( const item *e ) {
         if( e->get_quality( qual ) >= level ) {
-            qty += e->count_by_charges() ? e->charges : 1;
+            qty = sum_no_wrap( qty, e->count_by_charges() ? int( e->charges ) : 1 );
```

Comment:
sum_no_wrap() requires its arguments to be of the same type since it uses this to infer what value to use for wrapping, please don't ask for me to expand the template to handle arbitrary combinations of integer types, I won't do it.  This PR has already had feature creep dozens of times what it should have been.
Late narrowing is only sensical when dealing with varying precisions, with integer width it's either wide enough for the value or it isn't.  In this particular case we never expect to set item:charges to a value in excess of INT_MAX, so int is sufficient.  We are only using long values to deal with aggregations of item charge values.